cli/internal/oauth/manager/manager.go

```diff
@@ -0,0 +1,161 @@
+package manager
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/docker/cli/cli/internal/oauth/api"
+	"github.com/docker/cli/cli/internal/oauth/util"
+	"github.com/docker/cli/cli/oauth"
+)
+
+// OAuthManager is the manager
+type OAuthManager struct {
+	api         api.OAuthAPI
+	audience    string
+	tenant      string
+	openBrowser func(string) error
+}
+
+// OAuthManagerOptions is the options used for New to create a new auth manager.
+type OAuthManagerOptions struct {
+	Audience    string
+	ClientID    string
+	Scopes      []string
+	Tenant      string
+	DeviceName  string
+	OpenBrowser func(string) error
+}
+
+func New(options OAuthManagerOptions) *OAuthManager {
+	scopes := []string{"openid", "offline_access"}
+	if len(options.Scopes) > 0 {
+		scopes = options.Scopes
+	}
+
+	openBrowser := util.OpenBrowser
+	if options.OpenBrowser != nil {
+		openBrowser = options.OpenBrowser
+	}
+
+	return &OAuthManager{
+		audience: options.Audience,
+		api: api.API{
+			BaseURL:  "https://" + options.Tenant,
+			ClientID: options.ClientID,
+			Scopes:   scopes,
+			Client: util.Client{
+				UserAgent: options.DeviceName,
+			},
+		},
+		tenant:      options.Tenant,
+		openBrowser: openBrowser,
+	}
+}
+
+// LoginDevice launches the device authentication flow with the tenant, printing instructions
+// to the provided writer and attempting to open the browser for the user to authenticate.
+// Once complete, the retrieved tokens are stored and returned.
+func (m *OAuthManager) LoginDevice(ctx context.Context, w io.Writer) (res oauth.TokenResult, err error) {
+	state, err := m.api.GetDeviceCode(ctx, m.audience)
+	if err != nil {
+		return res, fmt.Errorf("login failed: %w", err)
+	}
+
+	if state.UserCode == "" {
+		return res, errors.New("login failed: no user code returned")
+	}
+
+	_, _ = fmt.Fprintln(w, "\nYou will be signed in using a web-based login.")
+	_, _ = fmt.Fprintln(w, "To sign in with credentials on the command line, use 'docker login -u <username>'")
+	_, _ = fmt.Fprintf(w, "\nYour one-time device confirmation code is: %s\n", state.UserCode)
+	_, _ = fmt.Fprint(w, "\nPress ENTER to open the browser.\n")
+	_, _ = fmt.Fprintf(w, "Or open the URL manually: %s.\n", strings.Split(state.VerificationURI, "?")[0])
```

Comment:
```suggestion
	_, _ = fmt.Fprintf(w, "Or open the URL manually: %s\n", strings.Split(state.VerificationURI, "?")[0])
```

My terminal emulator included the `.` when clicking on the link.

Reply:
Thanks! Fixed.